File: kranji-reading/kranji-reading-app/src/main/resources/homing/js/kranji/reading/app/known/KnownRecordModule.js
// =============================================================================
// KnownRecordModule — the known set, held as something you can ask.
//
// The set used to be a plain array, and `has` was indexOf. That is the fault
// underneath most of the update path: with no membership structure, every
// consumer that needed one built its own - readability per article, the
// toggle before it sent, the Secretary again to apply it. One record, asked
// rather than scanned, removes all of them at once.
//
// Two structures, deliberately. The map answers `has` in constant time; the
// array remembers insertion order, which is what makes an export file
// diffable and a list of claims read chronologically. Removal splices the
// array, which is linear - and stays linear, because a reader removes a claim
// far less often than the app asks whether one exists.
//
// The key is (character, reading). See KnownSetModule for why, and for keyOf,
// which builds it. This module never parses a key except to split off the
// codepoint for a character count.
//
// Pure - no DOM, no storage, no clock, no party.
// =============================================================================

/**
 * A record of claimed (character, reading) keys.
 *
 * `seed` is optional and is a union, never a replace, for the same reason the
 * Secretary's seed was: a record that arrives late - a slow disk, a service
 * that started after marking began - must not take back a claim made in the
 * meantime.
 *
 * Returns { has, add, remove, seed, keys, size, characters }.
 */
function createKnownRecord(seed) {

    var claimed = Object.create(null);
    var order   = [];

    /** True if this key was not already held. The answer callers act on. */
    function add(key) {
        if (!key || claimed[key]) return false;
        claimed[key] = true;
        order.push(key);
        return true;
    }

    /** True if this key was held. False means the caller asked for a no-op. */
    function remove(key) {
        if (!key || !claimed[key]) return false;
        delete claimed[key];
        var at = order.indexOf(key);
        if (at >= 0) order.splice(at, 1);
        return true;
    }

    function union(keys) {
        var added = 0;
        var list = keys || [];
        for (var i = 0; i < list.length; i++) if (add(list[i])) added++;
        return added;
    }

    union(seed);

    return {

        /** Constant time. This is the whole reason the module exists. */
        has: function (key) { return !!key && claimed[key] === true; },

        add:    add,
        remove: remove,

        /** A union, never a replace. Returns how many keys were new. */
        seed: union,

        /**
         * A copy, in insertion order. A copy because a caller that could
         * mutate this would be a second owner of the set, which is the thing
         * being removed.
         */
        keys: function () { return order.slice(); },

        size: function () { return order.length; },

        /**
         * Distinct characters, as codepoints.
         *
         * A character counts once however many of its readings are claimed,
         * which is why this is not size() and why "characters known" and
         * "readings known" must never be shown as the same number.
         */
        characters: function () {
            var seen = Object.create(null);
            var out  = [];
            for (var i = 0; i < order.length; i++) {
                var cp = order[i].slice(0, order[i].indexOf(':'));
                if (!seen[cp]) { seen[cp] = true; out.push(Number(cp)); }
            }
            return out;
        }
    };
}

File: kranji-reading/kranji-reading-app/src/main/java/kranji/reading/app/known/KnownRecordModule.java
package kranji.reading.app.known;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The known set, held as something you can ask rather than something you scan.
 *
 * <h2>Why this exists</h2>
 *
 * <p>The set was a plain array and {@code has} was {@code indexOf}. That is the
 * fault underneath most of the update path: with no membership structure, every
 * consumer that needed one built its own — readability rebuilt a claimed map
 * once per article, the reader's toggle scanned to decide direction, and the
 * Secretary scanned again to apply the result. Ranking the library at 2,000
 * readings known cost 23ms, almost all of it re-indexing the same set 475
 * times.</p>
 *
 * <p>One record, asked rather than scanned, removes all of them together.</p>
 *
 * <h2>Two structures</h2>
 *
 * <p>A map answers {@code has} in constant time; an array remembers insertion
 * order, which is what makes an export file diffable and a list of claims read
 * chronologically. Removal splices the array and so is linear — and stays
 * linear, because a reader takes a claim back far less often than the app asks
 * whether one exists.</p>
 *
 * <h2>Seeding is a union</h2>
 *
 * <p>Never a replace, for the reason the Secretary's seed was: a record that
 * arrives late — a slow disk, a service that started after marking began —
 * must not take back a claim made in the meantime.</p>
 *
 * <p>Pure — no DOM, no storage, no clock, no party — so it runs under GraalVM
 * in ordinary JUnit.</p>
 */
public record KnownRecordModule() implements DomModule<KnownRecordModule> {

    /** Yields {@code has}, {@code add}, {@code remove}, {@code seed} and the counts. */
    public record createKnownRecord() implements Exportable._Constant<KnownRecordModule> {}

    public static final KnownRecordModule INSTANCE = new KnownRecordModule();

    @Override
    public ImportsFor<KnownRecordModule> imports() {
        return ImportsFor.<KnownRecordModule>builder().build();
    }

    @Override
    public ExportsOf<KnownRecordModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new createKnownRecord()));
    }
}
